### PR TITLE
C++: Fix `hasDefaultSideEffect` for non-const smart pointers

### PR DESCRIPTION
**Pull request by MathiasVP**
Fixes https://github.com/github/codeql-c-team/issues/537.

CPP-Differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2059/

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/SideEffects.qll [hunk 1]
```diff
@@ -45,7 +55,7 @@ private predicate hasDefaultSideEffect(Call call, ParameterIndex i, boolean buff
       ) and
       (
         isWrite = true and
-        not call.getTarget().getParameter(i).getType().isDeeplyConstBelow()
+        not isConstPointerLike(call.getTarget().getParameter(i).getUnderlyingType())
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/SideEffects.qll:58]: Doesn't `isDeeplyConstBelow` require that, if you have multiple levels of indirection, that all levels are `const`? If so, that doesn't seem to be what `isConstPointerLike()` does.

**Review by dbartol — COMMENTED** [review]
One question.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/SideEffects.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/SideEffects.qll:58]: Hm, yes. Good point. I'll fix that (and probably add some more tests).

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/SideEffects.qll:106]: Fixed in 27586d77f87a51d42204745b8c7f1351cbf60131. I did a force push to make the commits come in a reviewer-friendly order.

Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/SideEffects.qll:106]: 👍

**Review by dbartol — APPROVED** [review]
LGTM

**Pull request merged**